widget/date_entry.go

```diff
@@ -0,0 +1,158 @@
+package widget
+
+import (
+	"time"
+
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/theme"
+)
+
+const dateFormat = "02 Jan 2006"
+
+// DateEntry is an input field which supports selecting from a fixed set of options.
+//
+// Since: 2.6
+type DateEntry struct {
+	Entry
+	Date      *time.Time
+	OnChanged func(*time.Time) `json:"-"`
+
+	dropDown *Calendar
+	popUp    *PopUp
+}
+
+// NewDateEntry creates a date input where the date can be selected or typed.
+//
+// Since: 2.6
+func NewDateEntry() *DateEntry {
+	e := &DateEntry{}
+	e.ExtendBaseWidget(e)
+	e.Wrapping = fyne.TextWrap(fyne.TextTruncateClip)
+	return e
+}
+
+// CreateRenderer returns a new renderer for this select entry.
+//
+// Implements: fyne.Widget
+func (e *DateEntry) CreateRenderer() fyne.WidgetRenderer {
+	e.ExtendBaseWidget(e)
+	e.Validator = func(in string) error {
+		_, err := time.Parse(dateFormat, in)
+		return err
+	}
+	e.Entry.OnChanged = func(in string) {
+		if in == "" {
+			e.Date = nil
+
+			if f := e.OnChanged; f != nil {
+				f(nil)
+			}
+		}
+		t, err := time.Parse(dateFormat, in)
+		if err != nil {
+			return
+		}
+
+		e.Date = &t
+
+		if f := e.OnChanged; f != nil {
+			f(&t)
+		}
+	}
+
+	if e.ActionItem == nil {
+		e.ActionItem = e.setupDropDown()
+		if e.Disabled() {
+			e.ActionItem.(fyne.Disableable).Disable()
+		}
+	}
+
+	return e.Entry.CreateRenderer()
+}
+
+// Enable this widget, updating any style or features appropriately.
+//
+// Implements: fyne.DisableableWidget
+func (e *DateEntry) Enable() {
+	if e.ActionItem != nil {
+		e.ActionItem.(fyne.Disableable).Enable()
+	}
+	e.Entry.Enable()
+}
+
+// Disable this widget so that it cannot be interacted with, updating any style appropriately.
+//
+// Implements: fyne.DisableableWidget
+func (e *DateEntry) Disable() {
+	if e.ActionItem != nil {
+		e.ActionItem.(fyne.Disableable).Disable()
+	}
+	e.Entry.Disable()
+}
+
+// MinSize returns the minimal size of the select entry.
+//
+// Implements: fyne.Widget
+func (e *DateEntry) MinSize() fyne.Size {
+	e.ExtendBaseWidget(e)
+	return e.Entry.MinSize()
+}
+
+// Move changes the relative position of the date entry.
+//
+// Implements: fyne.Widget
+func (e *DateEntry) Move(pos fyne.Position) {
+	e.Entry.Move(pos)
+	if e.popUp != nil {
+		e.popUp.Move(e.popUpPos())
+	}
+}
+
+// Resize changes the size of the date entry.
+//
+// Implements: fyne.Widget
+func (e *DateEntry) Resize(size fyne.Size) {
+	e.Entry.Resize(size)
+	if e.popUp != nil {
+		e.popUp.Resize(fyne.NewSize(size.Width, e.popUp.Size().Height))
+	}
+}
+
+func (e *DateEntry) SetDate(d *time.Time) {
+	if d == nil {
+		e.Date = nil
+		e.Entry.SetText("")
```

Comment:
Should there be a return after this line? (thanks staticcheck for detecting null dereference below)